Improve Package Spec cloning

nkolev92 opened the pull request:
## Bug
Fixes: None, but being done as part of https://github.com/NuGet/Home/issues/6065
Regression: No  
If Regression then when did it last work:   
If Regression then how are we preventing it in future:   

## Fix
Details: 
The Package Spec is cloned 2 times per project during one no-op restore. In VS that's done 3x.   
Each call takes 8-9ms. 
This change cuts it to ~1ms per call.

As mentioned in the comments earlier, I did not want to use existing APIs, because they did not imply deep cloning, while it's very important that the package spec is deep cloned. 

Some types like NuGetFramework and similar are immutable so those are/will not be cloned. 

The test subject is new MVC no-op, MVC clones 4x times, 2x2 for the tool and project restore each (tool clone takes way less ofc)

### With JIT baseline:
|Name | Inc % | Inc | Exc % | Exc |
|-- | -- | -- | -- | -- |
|nuget.projectmodel!DependencyGraphSpec.WithProjectClosure | 6.1 | 65 | 0.0 | 0 |
|+ nuget.projectmodel!NuGet.ProjectModel.PackageSpec.Clone() | 6.1 | 65 | 0.0 | 0 |
|+ nuget.projectmodel!JsonPackageSpecReader.GetPackageSpec | 3.0 | 32 | 0.0 | 0 |
|+ nuget.projectmodel!PackageSpecWriter.Write | 2.8 | 30 | 0.0 | 0 |
|+ clr!? | 0.3 | 3 | 0.0 | 0|

###  With JIT improved spec:

|Name | Inc % | Inc | Exc % | Exc|
|-- | -- | -- | -- | --|
|nuget.projectmodel!DependencyGraphSpec.WithProjectClosure | 0.8 | 8 | 0.0 | 0|
|+ nuget.projectmodel!NuGet.ProjectModel.PackageSpec.Clone() | 0.6 | 6 | 0.0 | 0|
|+ clr!? | 0.2 | 2 | 0.0 | 0|

### Without JIT baseline:

|Name | Inc % | Inc | Exc % | Exc|
|-- | -- | -- | -- | --|
|nuget.projectmodel!DependencyGraphSpec.WithProjectClosure | 1.5 | 9 | 0.0 | 0|
|+ nuget.projectmodel!NuGet.ProjectModel.PackageSpec.Clone() | 1.5 | 9 | 0.0 | 0|
|+ nuget.projectmodel!PackageSpecWriter.Write | 0.8 | 5 | 0.0 | 0|
|+ nuget.projectmodel!JsonPackageSpecReader.GetPackageSpec | 0.6 | 4 | 0.0 | 0|

###  WIthout JIT improved spec:

|Name | Inc % | Inc | Exc % | Exc|
|-- | -- | -- | -- | --|
|nuget.projectmodel!DependencyGraphSpec.WithProjectClosure | 0.2 | 1 | 0.0 | 0|
|+ nuget.projectmodel!NuGet.ProjectModel.PackageSpec.Clone() | 0.2 | 1 | 0.0 | 0|


Fixed an issue where RestoreArgs didn't work well with duplicate sources, because before this change the deduplication was totally accidentally done in the dg spec cloning [here](https://github.com/NuGet/NuGet.Client/blob/2c4e39c5c5cd52c7a2739fdcfd6bd2b212212cd1/src/NuGet.Core/NuGet.ProjectModel/PackageSpecWriter.cs#L256-L270)
More specifically the code change is this [one](https://github.com/NuGet/NuGet.Client/pull/1796/files#diff-4dc98f9946385b4b19fd5d9223d01fc0)

## Testing/Validation
Tests Added: Yes
Reason for not adding tests:  
Validation done:  Added new tests, existing automation tests, some manual validation

nkolev92 committed:
Initial cloning optimization work

nkolev92 committed:
More cloning owkr

nkolev92 committed:
add tests, clean up

nkolev92 committed:
fix tests

nkolev92 committed:
add mroe tests

nkolev92 committed:
Clone restore metadatA

nkolev92 committed:
Add more tests

nkolev92 committed:
Update PackageSpec.cs

nkolev92 committed:
revert header change

nkolev92 committed:
bla

nkolev92 committed:
add more tests

nkolev92 committed:
fix npe

nkolev92 committed:
cleanup

nkolev92 committed:
cleanup

nkolev92 committed:
cleanup

nkolev92 committed:
fix cloning

emgarten commented on src/NuGet.Core/NuGet.ProjectModel/IncludeExcludeFiles.cs:
```diff
@@ -86,5 +86,15 @@ public bool Equals(IncludeExcludeFiles other)
+
+        public IncludeExcludeFiles Clone()
+        {
+            var clonedObject = new IncludeExcludeFiles();
```
I don't think this is even used, maybe we could drop support for cloning some of these properties?

nkolev92 replied:
That could be the thing. 
I found a bunch of things that I don't see being used anymore in productive code, but actually being read in read/write.

emgarten replied:
🔪  they are all remnants of project.json, and things from DNU before this was ported to NuGet. Now is a good time to actually deprecate those.

jainaashish replied:
+1, it will also reduce the burden of packageSpec read wrtite which ultimately impact assets file read write as well.

emgarten commented:
Consider using the `ICloneable` interface https://msdn.microsoft.com/en-us/library/system.icloneable(v=vs.110).aspx

You could then make helper methods to clone lists of things without linq.

nkolev92 commented:
@emgarten I kind of opted against that ICloneable because it does not imply deep copy. 

Could consider it though.

emgarten commented:
@nkolev92 makes sense, if it doesn't do exactly when you need then you are better off skipping it. You could make a different interface for it, not sure if that would slow things down or make it easier.

nkolev92 commented:
@emgarten I think IDeepCloneable or something is an option. 

I'll look into it. 
Why do you think it would slow things down? Or in which way?

nkolev92 committed:
more tests

nkolev92 committed:
More tests

nkolev92 committed:
more tests

nkolev92 committed:
More tests

nkolev92 committed:
Fix tests..don't add duplicate package sources

nkolev92 committed:
don't do ToDictionary as it throws exception for duplicate soruces

nkolev92 committed:
revert v3 feed dedup

nkolev92 committed:
cleanup

nkolev92 committed:
Add more tests and runtime graph cloning

nkolev92 commented:
@emgarten 
Cleaned up a little bit, and added tests, so it's ready to review whenever :) 
Didn't make the Interface change, but if more people agree to that, I can change it. That would involve changing a bunch of older classes though...so not sure about the scope of that change.

emgarten commented on src/NuGet.Core/NuGet.Packaging/RuntimeModel/RuntimeGraph.cs:
```diff
@@ -50,7 +50,7 @@ private RuntimeGraph(Dictionary<string, RuntimeDescription> runtimes, Dictionary
         public RuntimeGraph Clone()
         {
-            return new RuntimeGraph(Runtimes.Values.Select(r => r.Clone()));
+            return new RuntimeGraph(Runtimes.Values.Select(r => r.Clone()), Supports.Values.Select( s => s.Clone()));
```
💯

emgarten commented on src/NuGet.Core/NuGet.ProjectModel/ProjectRestoreSettings.cs:
```diff
@@ -17,5 +17,37 @@ public class ProjectRestoreSettings
+
+        public override int GetHashCode()
+        {
+            return base.GetHashCode();
```
This should hash `HideWarningsAndErrors`

nkolev92 replied:
I autogened that...guess it I didn't include the field by accident. 

Updated it with another autogened hashcode.

emgarten commented on src/NuGet.Core/NuGet.ProjectModel/PackageSpec.cs:
```diff
@@ -185,24 +186,55 @@ public bool Equals(PackageSpec other)
-        public PackageSpec WithSettings(ISettings settings)
+        private IDictionary<string, IEnumerable<string>> CloneScripts(IDictionary<string, IEnumerable<string>> toBeCloned)
         {
+            if(toBeCloned != null)
```
nit: spacing

emgarten reviewed (changes requested):
Double check hash code for HideWarningsAndErrors

rohit21agrawal commented on src/NuGet.Core/NuGet.Packaging.Core/PackageType.cs:
```diff
@@ -1,11 +1,14 @@
 {
+    /**
+     * It is important that this type remains immutable due to the cloning of package specs
+     **/
```
have you done anything to enforce this? if this is the case, you should not have setters on any properties and all fields should be private

nkolev92 replied:
It doesn't have any setters, I just put it there, as it's one of the few types that are never cloned.

rohit21agrawal commented on src/NuGet.Core/NuGet.ProjectModel/PackOptions.cs:
```diff
@@ -46,5 +46,20 @@ public bool Equals(PackOptions other)
                    Mappings.SequenceEqualWithNullCheck(other.Mappings) &&
                    EqualityUtility.EqualsWithNullCheck(IncludeExcludeFiles, other.IncludeExcludeFiles);
         }
+        public PackOptions Clone()
```
PackOptions is not used anymore either. Remove it completely

rohit21agrawal reviewed (changes requested):
remove PackInclude, PackExclude and PackOptions completely from PackageSpec.

nkolev92 committed:
auto-generated hash code function project restore settings

nkolev92 committed:
spacing

nkolev92 commented:
@rohit21agrawal @emgarten 

Addressed the minor comments. 

PackInclude/PackOptions have lots of references though (PackCommandRunner mostly)
I'd rather not increase the scope of this change and create a different task for that, even if I do end up doing it myself. 

Do you guys agree?

emgarten commented on src/NuGet.Core/NuGet.ProjectModel/ProjectRestoreSettings.cs:
```diff
@@ -47,7 +47,7 @@ public bool Equals(ProjectRestoreSettings other)
         public override int GetHashCode()
         {
-            return base.GetHashCode();
+            return 318094734 + HideWarningsAndErrors.GetHashCode();
```
Use https://github.com/NuGet/NuGet.Client/blob/dev/src/NuGet.Core/NuGet.Commands/Internal/HashCodeCombiner.cs 

It is included in all projects

nkolev92 replied:
Done.

nkolev92 committed:
use hash code combiner

emgarten commented on src/NuGet.Core/NuGet.ProjectModel/ProjectRestoreSettings.cs:
```diff
@@ -17,5 +18,39 @@ public class ProjectRestoreSettings
+        public override int GetHashCode()
+        {
+            var hashCode = new HashCodeCombiner();
+            hashCode.AddObject(HideWarningsAndErrors);
```
👍

emgarten reviewed (approved):
Fine with me if the removal is done later. I think PackageSpec just needs some good clean up, and that isn't directly related to perf.

nkolev92 commented:
Cool...created https://github.com/NuGet/Home/issues/6231. 

Didn't assign it any priority yet, hopefully we can get it done within the next few weeks.

jainaashish commented on src/NuGet.Core/NuGet.Commands/RestoreCommand/RequestFactory/RestoreArgs.cs:
```diff
@@ -128,9 +128,13 @@ internal List<SourceRepository> GetEffectiveSources(ISettings settings, IList<Pa
             var packageSourceProvider = new PackageSourceProvider(settings);
             var packageSourcesFromProvider = packageSourceProvider.LoadPackageSources();
+            var sourceObjects = new Dictionary<string, PackageSource>();
+            foreach(var source in dgSpecSources)
```
nit: since `dgSpecSources` is a list then you can further improve it by replacing `ForEach` with `For`. Which will avoid `Enumeration` altogether from this. It might not make much difference, but just a good perf practice

nkolev92 replied:
That's a good point. 
It doesn't help us much, but it's a good practice.

mishra14 commented on src/NuGet.Core/NuGet.Packaging/RuntimeModel/RuntimeGraph.cs:
```diff
@@ -50,7 +50,7 @@ private RuntimeGraph(Dictionary<string, RuntimeDescription> runtimes, Dictionary
         public RuntimeGraph Clone()
         {
-            return new RuntimeGraph(Runtimes.Values.Select(r => r.Clone()));
+            return new RuntimeGraph(Runtimes.Values.Select(r => r.Clone()), Supports.Values.Select( s => s.Clone()));
```
Should this also clone `_areCompatible`? Not cloning that saves us time now, but may have runtime cost.

nkolev92 replied:
That's not any different from how it is/was before this change.
That's not a hot path anyways, so I'd rather not clone concurrent dictionaries around.

jainaashish commented on src/NuGet.Core/NuGet.ProjectModel/PackageSpec.cs:
```diff
@@ -185,24 +186,55 @@ public bool Equals(PackageSpec other)
+            spec.Summary = Summary;
+            spec.ReleaseNotes = ReleaseNotes;
+            spec.Authors = (string[]) Authors?.Clone();
+            spec.Owners = (string[]) Owners?.Clone();
```
why do you need to clone these Authors or Owners?

nkolev92 replied:
Why shouldn't they be cloned? 
They are a part of the model and they're mutable. Despite the fact that no one changes them, the clone method should account for that.

mishra14 commented on src/NuGet.Core/NuGet.ProjectModel/BuildOptions.cs:
```diff
@@ -38,5 +38,12 @@ public bool Equals(BuildOptions other)
+
+        public BuildOptions Clone()
+        {
+            var options = new BuildOptions();
```
you should be able to simplify this to - 

```
return new BuildOptions()
{
    OutputName = this.OutputName
};
```

mishra14 replied:
Also keep the naming consistent to `clonedObject`.

mishra14 commented on src/NuGet.Core/NuGet.ProjectModel/PackageSpec.cs:
```diff
@@ -185,24 +186,55 @@ public bool Equals(PackageSpec other)
+            }
             return null;
         }
+
```
nit: extra line.

jainaashish reviewed (approved):
Few minor comments, and also please get rid of unused code related to this cloning.

nkolev92 committed:
cleanup

mishra14 reviewed (approved):
small feedback

nkolev92 committed:
address feedback
